File: com.pck.binary/FirstAndLast.java
class Solution {
    // Time complexity: O(logn)
    // Space Complexity :O(1)  
    public int[] searchRange(int[] nums, int target) {
        int first = binarySearchFirst(nums, target);
        if(first == -1){
            return new int[]{-1, -1};
        }
        int second = binarySearchSecond(nums, target);
        return new int[]{first, second};
    }

    private int binarySearchFirst(int[] nums, int target){
        int low = 0;
        int high = nums.length - 1;
        while(low <= high){
            int mid =  low + (high - low) / 2;
            if(nums[mid] == target){
                if(mid == low || nums[mid - 1] < nums[mid]){
                    return mid;
                } else {
                    high = mid - 1;
                }

            } else if(nums[mid] < target){
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return -1;
    }

    private int binarySearchSecond(int[] nums, int target){
        int low = 0;
        int high = nums.length - 1;
        while(low <= high){
            int mid =  low + (high - low) / 2;
            if(nums[mid] == target){
                if(mid == high || nums[mid + 1] > nums[mid]){
                    return mid;
                } else {
                    low = mid + 1;
                }

            } else if(nums[mid] < target){
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return -1;
    }
}

File: first_and_last.py
# time: o(log n)
# space: O(log n)
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        lower = self.binSearch(nums, 0, len(nums)-1, target, True)
        if lower == -1:
            return [-1,-1]
        upper = self.binSearch(nums, 0, len(nums)-1, target, False)
        return [lower, upper]

    def binSearch(self, nums, l,r, target, isFirst):
        while l<=r:
            m = (l+r)//2
            if target == nums[m]:
                if isFirst:
                    if m==l or nums[m]>nums[m-1]:
                        return m
                    else:
                        r=r-1
                else:
                    if m==r or nums[m]<nums[m+1]:
                        return m
                    else:
                        l=l+1
            elif target > nums[m] :
                l = m+1
            else:
                r = m-1

        return -1
                 


        